### PR TITLE
Feature/#136 ZIP import

### PR DESCRIPTION
**Pull request by kaklakariada**
See #136

### CODE REVIEW

**Comment by kaklakariada** [comment]
@redcatbear This is not yet finished yet but refactoring (using `InputFile` as abstraction layer for importer) is done. Please review the refactoring.

**Comment by kaklakariada** [comment]
@redcatbear Implementation finished, please review.

File: src/main/java/org/itsallcode/openfasttrace/exporter/ExporterFactory.java [hunk 1]
```diff
@@ -93,12 +89,19 @@ private Writer createWriter(final Path file, final Charset charset)
         if (file == null)
         {
             LOG.finest(() -> "Creating exporter for stdout using charset " + charset);
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/exporter/ExporterFactory.java]: STDOUT

File: src/main/java/org/itsallcode/openfasttrace/exporter/ExporterFactory.java [hunk 2]
```diff
@@ -93,12 +89,19 @@ private Writer createWriter(final Path file, final Charset charset)
...
-            return new OutputStreamWriter(System.out, charset);
+            return new OutputStreamWriter(getStdOutStream(), charset);
         }
         LOG.finest(() -> "Creating exporter for file " + file + " using charset " + charset);
         return createFileWriter(file, charset);
     }
 
+    // Using System.out by intention
+    @SuppressWarnings("squid:S106")
+    private PrintStream getStdOutStream()
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/exporter/ExporterFactory.java]: :+1:

File: src/main/java/org/itsallcode/openfasttrace/importer/ImporterContext.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+import java.util.Objects;
+
+import org.itsallcode.openfasttrace.importer.legacytag.config.LegacyTagImporterConfig;
+
+public class ImporterContext
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/ImporterContext.java]: Please document all public classes and methods.

File: src/main/java/org/itsallcode/openfasttrace/importer/ImporterFactory.java [hunk 4]
```diff
@@ -1,72 +1,73 @@
...
-        {
-            return Files.newBufferedReader(file, charset);
-        }
-        catch (final IOException e)
-        {
-            throw new ImporterException("Error reading file '" + file + "': " + e.getMessage(), e);
-        }
-    }
-}
+package org.itsallcode.openfasttrace.importer;
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/ImporterFactory.java]: Move all imports below the license.

File: src/main/java/org/itsallcode/openfasttrace/importer/ImporterContext.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+{
+    private final LegacyTagImporterConfig tagImporterConfig;
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/ImporterContext.java]: This looks like unintended information disclosure to me.
If all importers get this context, all have access to the `tagImporterConfig` - even if they are not the `TagImporter`.

There must be a better way where only the `TagImporter` learns about its config.

File: src/main/java/org/itsallcode/openfasttrace/importer/ImporterFactoryLoader.java [hunk 6]
```diff
@@ -22,7 +22,6 @@
  * #L%
  */
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/ImporterFactoryLoader.java]: Move package below license.

File: src/main/java/org/itsallcode/openfasttrace/importer/input/InputFile.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+public interface InputFile
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/input/InputFile.java]: Please document public class and methods.

File: src/main/java/org/itsallcode/openfasttrace/importer/ImporterContext.java (hunk 5 above)
Comment (kaklakariada) [reply src/main/java/org/itsallcode/openfasttrace/importer/ImporterContext.java]: I agree. I propose to address this in #144

**Comment by redcatbear** [comment]
Please remember to add JavaDoc to the remaining public methods and classes.

**Pull request merged**